grudge/symbolic/operators.py

```diff
@@ -651,20 +658,31 @@ def matrix(self, afgrp, volgrp, dtype):
                     afgrp.nunit_dofs),
                 dtype=dtype)
 
-        from modepy.tools import UNIT_VERTICES
         import modepy as mp
-        for iface, fvi in enumerate(
-                volgrp.mesh_el_group.face_vertex_indices()):
-            face_vertices = UNIT_VERTICES[volgrp.dim][np.array(fvi)].T
-            matrix[:, iface, :] = mp.nodal_face_mass_matrix(
-                    volgrp.basis(), volgrp.unit_nodes, afgrp.unit_nodes,
-                    volgrp.order,
-                    face_vertices)
-
-        # np.set_printoptions(linewidth=200, precision=3)
-        # matrix[np.abs(matrix) < 1e-13] = 0
-        # print(matrix)
-        # 1/0
+
+        vol_basis = volgrp.basis_obj()
+        faces = mp.faces_for_shape(volgrp.shape)
+
+        for iface, face in enumerate(faces):
+            # NOTE: Quadrature degree determined such that
+            # we either guarantee exact integration of degree 2N
+            # polynomials, or use a higher-order quadrature rule
+            # provided by the user (for example, when doing overintegration)
```

Comment:
I don't understand the motivation here. Why is the volume being treated differently than the surface?

Reply:
@inducer When you overintegrate, the `afgrp` is coming from the higher-order quadrature grid. This is just a note for why I computed `face_quad_degree` below. If I just took the order from the volume group (i.e. `face_quad_degree = 2*volgrp.order`, this causes the diffusion test in mirgecom to fail (because it's actually using overintegration.

Take a look at this test here to see the example I'm trying to handle `https://github.com/illinois-ceesd/mirgecom/blob/main/test/test_diffusion.py#L271`

If you think this can be handled in a different way, I'm open to suggestions.

Reply:
Thanks for poking at this. It turns out that it was much more broken than I thought, cf. https://github.com/inducer/modepy/pull/28. There should genuinely be two cases here, nodal and overintegrated.

Reply:
(So, to be clear, I'm proposing that we base the work here on the new functionality in https://github.com/inducer/modepy/pull/28, so we'll need a modepy version lower bound here.)